### ISSUE TITLE
Consider the snoop library

### ISSUE THREAD

**Issue by alexmojaki** [zasdfgbnm/TorchSnooper#15]
I've just released [snoop](https://github.com/alexmojaki/snoop), an improved version of PySnooper. It already shows the length and shape of all values, and it's easy to add similar automatic watches. For example:

```python
import torch
import snoop


def watch_device(source, value):
    return '{}.device'.format(source), value.device


@snoop(watch_extras=[watch_device])
def myfunc():
    y = torch.zeros(6)
    y = y.reshape((2, 3))
    return y


myfunc()
```

Output:

![Screen Shot 2019-07-05 at 21 02 36](https://user-images.githubusercontent.com/3627481/60741045-5b60da80-9f68-11e9-9d37-1019267ce710.png)

Note that:

1. The original repr still shows, so users can see both the contents and the metadata. And unlike PySnooper, newlines are not removed.
2. This works for anything with a `device` attribute, such as proxy objects.
3. It doesn't show all the metadata every time. For example, when the device doesn't change, it doesn't show it again.

Alternatively you could register a function in `cheap_repr`:

```python
import torch
import snoop
from cheap_repr import register_repr, cheap_repr


@register_repr(type(torch.zeros(6)))
def repr_torch_tensor(x, _helper):
    return 'tensor(shape={}, device={}, dtype={})'.format(
        cheap_repr(x.shape),
        x.device,
        str(x.dtype)[len('torch.'):],
    )


@snoop
def myfunc():
    y = torch.zeros(6)
    y = y.reshape((2, 3))
    return [{'key': y}]


myfunc()
```

Output:

![Screen Shot 2019-07-05 at 21 13 23](https://user-images.githubusercontent.com/3627481/60741467-c828a480-9f69-11e9-8d6f-2e5f5708186b.png)

This has the advantage of also showing the repr in containers as well as other classes, so you don't have to worry about that yourself.

Let me know what seems best or if you have other suggestions.

**Comment by zasdfgbnm** [comment]
Hi @alexmojaki ,

Thanks for your information and for developing snoop! I do like the UI of snoop more than PySnooper, but I don't want to push my user to switch. But alternatively, it would be convenient to add support for snoop. Basically, I hope torchsnooper would looks like follows:

```python
import torch
import snoop
import torchsnooper


torchsnooper.register_snoop()
# or equivalently:
# torchsnooper.register_snoop(verbose=False)

@snoop
def myfunc():
    y = torch.randn(6, 6, 6)
    return y

myfunc()
```
In this case, we should have the output

```
18:05:58.28 >>> Call to myfunc in File "test.py", line 9
18:05:58.28    9 | def myfunc():
18:05:58.28   10 |     y = torch.randn(6, 6, 6)
18:05:58.28 .......... y = tensor<(6, 6, 6), float32, cpu>
18:05:58.28   11 |     return y
18:05:58.28 <<< Return value from myfunc: tensor<(6, 6, 6), float32, cpu>
```
Note that there is no extra `len(y)` and `y.shape` line displayed for pytorch tensor in terse mode.

Alternatively, the user could choose to enable verbose mode, by changing the 
```python
torchsnooper.register_snoop()
```
into
```python
torchsnooper.register_snoop(verbose=True)
```
in this case, we should expect
```
18:09:54.52 >>> Call to myfunc in File "test.py", line 9
18:09:54.52    9 | def myfunc():
18:09:54.52   10 |     y = torch.randn(6, 6, 6)
18:09:54.52 .......... y = tensor([[[ 1.0272e+00,  1.7453e-01,  6.3193e-01,....0854e+00,  1.1385e-01,
18:09:54.52                          -9.7522e-01]]])
18:09:54.52 .......... y.shape = (6, 6, 6)
18:09:54.52 .......... y.dtype = float32
18:09:54.52 .......... y.device = cpu
18:09:54.52   11 |     return y
18:09:54.52 <<< Return value from myfunc: tensor([[[ 1.0272e+00,  1.7453e-01,  6.3193e-01,....0854e+00,  1.1385e-01,
18:09:54.52                                         -9.7522e-01]]])
```
Note that there is no `len(y)` line.

Please let me know your opinion, and I do have a few questions:
1. How do I "register" a watch, only for a specific type, so that the user don't need to manually specify it inside `@snoop`?
2. How do I cancel a watch that was builtin by default (here the `len` and `shape` watch)

**Comment by alexmojaki** [comment]
Awesome! I think the idea for your API is good, it means people can continue to use snoop as normal instead of deciding between @snoop and @torchsnooper.snoop.

I've moved the `watch_extras` argument to `install` so that it's a global configuration. It's something I suspected I needed to do anyway. In the case of this library, you wouldn't want to interfere with the user's `install()`, so you would write:

```python
import snoop

def register_snoop():
    ...
    snoop.config.watch_extras += (device_watch,)
```

and tell the user to call `register_snoop` after `install()`.

I've changed the default watches so that:

1. If something has a shape, it only shows that and not the `len()`. I think you're right that showing both is generally pointless.
2. I've added a `.dtype` watch to the defaults, since it's so common.

I've pushed these changes to master but not to PyPI, don't want to be making more releases in a rush.
